ouroboros-consensus/test/storage-test/Test/Ouroboros/Storage/LedgerDB/V1/BackingStore/Mock.hs

```diff
@@ -0,0 +1,338 @@
+      copies = Map.insert bsp (backingSeqNo m, backingValues m) (copies m)
+    })
+
+-- | Open a new value handle, which captures the state of the backing store
+-- at the time of opening the handle.
+mBSValueHandle ::
+     (MonadState (Mock vs) m, MonadError Err m)
+  => m (ValueHandle vs)
+mBSValueHandle = do
+  mGuardBSClosed
+  vs <- gets backingValues
+  seqNo <- gets backingSeqNo
+  nxt <- gets nextId
+  let
+    vh = ValueHandle nxt vs seqNo
+  modify (\m -> m {
+      valueHandles = Map.insert nxt Open (valueHandles m)
+    , nextId = nxt + 1
+    })
+
+  pure vh
+
+-- | Write a diff to the backing store.
+mBSWrite ::
+     (MonadState (Mock vs) m, MonadError Err m, ApplyDiff vs d)
+  => SlotNo
+  -> d
+  -> m ()
+mBSWrite sl d = do
+  mGuardBSClosed
+  vs <- gets backingValues
+  seqNo <- gets backingSeqNo
+  when (seqNo > NotOrigin sl) $
+    throwError $ ErrNonMonotonicSeqNo (NotOrigin sl) seqNo
+  modify (\m -> m {
+      backingValues = applyDiff vs d
+    , backingSeqNo = NotOrigin sl
+    })
+
+-- | Throw an error if the given backing store value handle has been closed.
+mGuardBSVHClosed ::
+     (MonadState (Mock vs) m, MonadError Err m)
+  => ValueHandle vs
+  -> m ()
+mGuardBSVHClosed vh = do
+  vhs <- gets valueHandles
+  case Map.lookup (getId vh) vhs of
+    Nothing -> error "Value handle not found"
+    Just status ->
+      case status of
+        ClosedByStore  -> throwError ErrBackingStoreClosed
+        ClosedByHandle -> throwError ErrBackingStoreValueHandleClosed
+        _              -> pure ()
+
+-- | Close a backing store value handle.
+--
+-- Closing is idempotent.
+mBSVHClose ::
+     (MonadState (Mock vs) m, MonadError Err m)
+  => ValueHandle vs
+  -> m ()
+mBSVHClose vh =
+    (mGuardBSClosed >> mGuardBSVHClosed vh >> close) `catchError` handler
+  where
+    close = do
+      vhs <- gets valueHandles
+      modify (\m -> m {
+          valueHandles = Map.adjust (const ClosedByHandle) (getId vh) vhs
+        })
+    handler = \case
+      ErrBackingStoreClosed            -> pure ()
+      ErrBackingStoreValueHandleClosed -> pure ()
+      e                                -> throwError e
+
+-- | Perform a range read on a backing store value handle.
+mBSVHRangeRead ::
+     (MonadState (Mock vs) m, MonadError Err m, LookupKeysRange ks vs)
+  => ValueHandle vs
+  -> BS.RangeQuery ks
+  -> m vs
+mBSVHRangeRead vh BS.RangeQuery{BS.rqPrev, BS.rqCount} = do
+  mGuardBSClosed
+  mGuardBSVHClosed vh
+  let
+    vs = values vh
+  pure $ lookupKeysRange rqPrev rqCount vs
+
+-- | Perform a regular read on a backing store value handle
+mBSVHRead ::
+     (MonadState (Mock vs) m, MonadError Err m, LookupKeys ks vs)
+  => ValueHandle vs
+  -> ks
+  -> m vs
+mBSVHRead vh ks = do
+  mGuardBSClosed
+  mGuardBSVHClosed vh
+  let vs = values vh
+  pure $ lookupKeys ks vs
+
+-- | Read the slot number out of a value handle
+mBSVHAtSlot :: Monad m => ValueHandle vs -> m (WithOrigin SlotNo)
+mBSVHAtSlot = pure . seqNo
```

Comment:
Why not check if the backing store is closed? Seqno is stored in the LMDB database, isn't it?

Reply:
This slot number is the slot at which the value handle was created, not the current slot of the backing store. `bsvhAtSlot` is just a value, not an action